text/3748-cargo_check_environment_variable.md

```diff
@@ -0,0 +1,107 @@
+- Feature Name: `cargo_check_environment_variable`
+- Start Date: 2024-12-20
+- RFC PR: [rust-lang/rfcs#3748](https://github.com/rust-lang/rfcs/pull/3748)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new environment variable `CARGO_CHECK` that is set to `true` when running `cargo check` or similar type-checking operations so build scripts can skip expensive compilation steps that are unnecessary for Rust type checking, such as compiling external C++ code in cxx based projects.
+
+# Motivation
+[motivation]: #motivation
+
+Rust development heavily relies on IDE tooling like rust-analyzer, which frequently invokes `cargo check` to provide real-time type information and diagnostics. Many projects use build scripts (`build.rs`) to generate Rust code and compile external dependencies. For example:
+
+- cxx-rs generates Rust bindings for C++ code and compiles C++ source files
+- cxx-qt generates Rust bindings for Qt code and runs the Qt Meta-Object Compiler (MOC)
+- Projects using Protocol Buffers generate Rust code from .proto files
+- bindgen generates Rust bindings from C/C++ headers
+
+Currently, every time rust-analyzer runs `cargo check`, all build scripts must execute their full build process, including steps like compiling C++ code that are only needed for linking but not for type checking. This significantly impacts IDE responsiveness, especially in projects with complex build scripts.
```

Comment:
Something seems missing from this description. In my understanding of how to use Cargo, a well-behaved build script should produce an appropriate set of [`rerun-if-changed`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-changed) directives, meaning that the build script will not be run “every time” unless its dependencies are actually changed.

There are of course specific scenarios where it will be rerun regardless, such as if you are editing one of the `build-dependencies`, but this paragraph suggests that build scripts predictably run on *every* `check` in every project that has one, not just the first `cargo check` or `cargo build`. In my mind, if that’s happening, that indicates a bug in the build script. (Or at least an incompleteness — rerunning every time is the *default* behavior of Cargo given zero `rerun-if-*` directives, but for any build script that runs long enough to notice, that should be addressed by adding those directives.)

Can you clarify this? Is the build script in this scenario _unable_ to use `rerun-if` directives for some reason?

Reply:
You are correct, I miswrote that paragraph. I changed it to `Currently, every time rust-analyzer runs `cargo check`, the build script in the changed crate must execute its full build process, ...`

Reply:
But that’s still not true in general. If a build script reruns, then one of the following situations applies:

* The build script does not have any `rerun-if-*` dependencies declared; that can and should be fixed.
* You modified one of its `[build-dependencies]`.
* You modified a file that is a `rerun-if-changed` dependency of the build script, so it *should* rerun.
* You (implicitly or explicitly) changed a environment variable the build script depends on with `rerun-if-env-changed`, e.g. by running `cargo check` from two different environments (IDE and separate terminal). Again, this *is* an actual dependency, though possibly an unintended one.

Changing the Rust code of a crate *does not* guarantee that its build script will rerun. Build scripts are rerun only if one of their dependencies changes, or if they failed to declare dependencies.

Reply:
Ah I see what you mean now.

The `cxx` crate parses Rust source files to generate bindings from `cxx::bridge` macros and the Rust source files get added via `cargo::rerun-if-changed`. Normally this wouldn't be a problem since C/C++ bindings are rarely changed, but in `cxx-qt` projects the bridge code is the interface between QML/JS GUI code and the Rust backend so it receives a lot more attention while incurring additional QT build tooling overhead.

I'll add a paragraph explaining that detail. I didn't realize the issue I was experiencing is so specific to `cxx` though